Feature/2.x/194 add entity and bundle template suggestions to page header block

ctorgalson opened the pull request:
Re: https://github.com/localgovdrupal/localgov_core/issues/194

## What does this change?

This makes three non-breaking changes to `localgov_core`'s Page Header Block:

1. adds `#entity` to `localgov_page_header_block` in `localgov_core_theme()`
2. adds the current entity to `PageHeaderBlock.php`.s render array
3. adds Twig template suggestions to the block via `localgov_core_theme_suggestions_localgov_page_header_block()`

## How to test

There are five things requiring testing:

1. that _nothing changes_ on existing sites like https://demo.localgovdrupal.org
2. that three types of Twig template suggestions are provided:
  - per entity id, e.g. `localgov-page-header-block--node--7.html.twig`
  - per bundle, e.g. `localgov-page-header-block--localgov-services-landing.html.twig`
  - per entity type, e.g. `localgov-page-header-block--node.html.twig`
  
    ![image](https://github.com/user-attachments/assets/918bbe28-99f7-49e5-b0ad-bf1c075d6df9)
3. that the expected entity is present in Twig; add the following to `localgov-page-header-block.html.twig` (or one of the suggested ones) and verify on the front end:
    ```twig
    {#
      On e.g. a localgov_services_landing node, with nid 7, this should produce the following output:
      
    0 array:3 [
      0 => "7"
      1 => "localgov_services_landing"
      2 => "node"
    ]
    
    #}
    {{ dump([
      entity.id(),
      entity.bundle(),
      entity.getEntityTypeId(),
    ]) }}
    ```
4. that templates named according to the suggestions are discovered and rendered
5. that non-entity pages (like Views) don't break because of these changes

## How can we measure success?

Three ways:

1. if nothing whatever changes on any existing site
2. that it becomes possible to access the base entity from within the block template
3. that it becomes possible to override the module's template (and `localgov_base`'s template) with nid, bundle, and entity type specific templates.

## Have we considered potential risks?

As this is only intended to expand what's _possible_ without actually changing anything, it should be low risk (mainly the risk is that I've failed to grasp something important in the block plugin :smiley:)

## Images

n/a

## Accessibility

n/a

ctorgalson committed:
feat: adds current entity to block render array

ctorgalson committed:
feat: adds entity variable to page header block in localgov_core_theme()

ctorgalson committed:
feat: adds hook_theme_suggestions_HOOK() implementation

Provides two new suggestions:

- localgov_page_header_block__$entity->bundle()
- localgov_page_header_block__$entity->id()

ctorgalson committed:
feat: adds an entity type template suggestion

ctorgalson committed:
feat: adds return type to theme suggestions hook

ctorgalson committed:
feat: adds guard to ensure we CAN create entity-based suggestions

Now Views don't break :)

ctorgalson committed:
feat: adds NodeInterface, TermInterface 'use' statements

We check instanceof with each of these in the module file.

ctorgalson commented:
Looks like the still-failing test is not related to this change.

millnut commented on localgov_core.module:
```diff
@@ -1,5 +1,8 @@
 <?php
 
+use Drupal\node\NodeInterface;
+use Drupal\taxonomy\TermInterface;
+
```
Hi @ctorgalson the use statements need to go below the file comment, could you add these line below the `use` statement on line 11 and see if that gets the tests to continue

ctorgalson replied:
D'oh! Thanks for spotting that.

ctorgalson committed:
feat: corrects position of 'use' statements

ctorgalson committed:
feat: modifies id suggestion to avoid collisions

As these suggestions apply to the block on more than one entity type, we
could see the same id value on different entities. This modifies the id
suggestion to include the entity type, e.g.:

- localgov-page-header-block--node--7.html.twig
- localgov-page-header-block--term--7.html.twig

finnlewis commented:
Thanks @ctorgalson ! Let's try to test and review this week.

finnlewis commented:
@rupertj you might be interested in this one too!

markconroy reviewed (approved):
Works a treat, no impact on views pages.
